# 📐 Les Bases des Mathématiques

**Bienvenue dans ton voyage mathématique !** 🎉

---

## ⚠️ Prérequis : Environnement Python

**Avant de commencer ce cours, assure-toi d'avoir :**

1. ✅ Installé Miniconda (voir `cours/Phase_0_Fondations/01_Installation_Environnement.ipynb`)
2. ✅ Créé et activé l'environnement `llm`

**Si tu n'as pas encore créé l'environnement, ouvre un Terminal et tape :**

```bash
# 1. Créer l'environnement
conda create -n llm python=3.11 numpy pandas matplotlib seaborn jupyter ipykernel -y

# 2. Activer l'environnement
conda activate llm

# 3. Enregistrer dans Jupyter
python -m ipykernel install --user --name=llm --display-name="Python (LLM)"

# 4. Relancer Jupyter avec le bon kernel
jupyter notebook
```

**Puis sélectionne le kernel "Python (LLM)"** dans Jupyter (en haut à droite).

---

Tu connais déjà les fractions et les multiplications ? Parfait ! C'est tout ce qu'il te faut pour commencer. Dans ce cours, nous allons construire ensemble tes connaissances mathématiques, étape par étape, sans rien précipiter.

**Ce que tu vas apprendre :**
- Les différents types de nombres
- Les puissances (beaucoup plus simples que ça en a l'air !)
- Les racines (le contraire des puissances)
- Les variables (comme des boîtes magiques)
- Comment résoudre des équations simples

**Important :** Prends ton temps, teste les exemples en Python, et n'hésite pas à revenir en arrière si besoin. Les maths, c'est comme un jeu vidéo : on débloque les niveaux un par un !

---
## 1️⃣ Les Nombres : Notre Boîte à Outils

Imagine que les nombres sont comme différents types d'outils dans une boîte. Chaque type a son utilité !

### 🔢 Les Nombres Naturels

Ce sont les nombres que tu utilises pour compter : **0, 1, 2, 3, 4, 5...**

On les note : $$\mathbb{N} = \{0, 1, 2, 3, 4, 5, ...\}$$

**Pourquoi c'est utile ?** Pour compter des objets : "J'ai 3 pommes", "Il y a 15 élèves".

In [2]:
# Exemples de nombres naturels
pommes = 3
eleves = 15

print(f"J'ai {pommes} pommes")
print(f"Il y a {eleves} élèves")

# En Python, on peut vérifier si un nombre est naturel
nombre = 5
est_naturel = nombre >= 0 and isinstance(nombre, int)
print(f"{nombre} est un nombre naturel : {est_naturel}")

J'ai 3 pommes
Il y a 15 élèves
5 est un nombre naturel : True


### ➕➖ Les Nombres Entiers

Les nombres naturels **+ les nombres négatifs** : **..., -3, -2, -1, 0, 1, 2, 3, ...**

On les note : $$\mathbb{Z} = \{..., -3, -2, -1, 0, 1, 2, 3, ...\}$$

**Pourquoi c'est utile ?** Pour représenter des températures, des dettes, des niveaux (sous-sol : -1, rez-de-chaussée : 0, étage : 1).

In [3]:
# Exemples de nombres entiers
temperature_hiver = -5  # Il fait -5°C
argent_compte = -20     # Je dois 20€ (dette)
niveau_parking = -2     # Parking au sous-sol 2

print(f"Température : {temperature_hiver}°C")
print(f"Argent dans mon compte : {argent_compte}€")
print(f"Niveau de parking : {niveau_parking}")

# Tous les nombres naturels sont aussi des entiers
print(f"\n5 est un nombre naturel ET un nombre entier")

Température : -5°C
Argent dans mon compte : -20€
Niveau de parking : -2

5 est un nombre naturel ET un nombre entier


### 🍕 Les Nombres Rationnels (Fractions !)

Ce sont les nombres qu'on peut écrire comme **une fraction** : $$\frac{a}{b}$$ où $$a$$ et $$b$$ sont des entiers (et $$b \neq 0$$).

On les note : $$\mathbb{Q} = \left\{\frac{a}{b} \mid a, b \in \mathbb{Z}, b \neq 0\right\}$$

**Exemples :** $$\frac{1}{2}$$, $$\frac{3}{4}$$, $$\frac{-5}{2}$$, $$2$$ (car $$2 = \frac{2}{1}$$)

**Pourquoi c'est utile ?** Pour partager des choses : "Je mange la moitié d'une pizza" = $$\frac{1}{2}$$

In [4]:
# En Python, on peut utiliser fractions ou décimales
from fractions import Fraction

# Fraction (exact)
moitie = Fraction(1, 2)
trois_quarts = Fraction(3, 4)

print(f"Moitié : {moitie}")
print(f"Trois quarts : {trois_quarts}")

# Décimale (approximation)
moitie_decimal = 1 / 2
trois_quarts_decimal = 3 / 4

print(f"\nMoitié en décimal : {moitie_decimal}")
print(f"Trois quarts en décimal : {trois_quarts_decimal}")

# Addition de fractions
resultat = moitie + trois_quarts
print(f"\n{moitie} + {trois_quarts} = {resultat}")

Moitié : 1/2
Trois quarts : 3/4

Moitié en décimal : 0.5
Trois quarts en décimal : 0.75

1/2 + 3/4 = 5/4


### 🌊 Les Nombres Réels

Ce sont **TOUS les nombres** sur la droite des nombres, y compris ceux qu'on ne peut PAS écrire en fraction.

On les note : $$\mathbb{R}$$

**Exemples de nombres réels non-rationnels :**
- $$\pi = 3.14159...$$
- $$\sqrt{2} = 1.41421...$$
- $$e = 2.71828...$$

**Pourquoi c'est utile ?** Pour mesurer avec précision : la circonférence d'un cercle, la diagonale d'un carré...

In [5]:
import math

# Quelques nombres réels célèbres
pi = math.pi
racine_de_2 = math.sqrt(2)
e = math.e

print(f"π (pi) ≈ {pi}")
print(f"√2 (racine de 2) ≈ {racine_de_2}")
print(f"e (nombre d'Euler) ≈ {e}")

# Ces nombres ont une infinité de décimales sans répétition
print(f"\nπ avec 15 décimales : {pi:.15f}")

π (pi) ≈ 3.141592653589793
√2 (racine de 2) ≈ 1.4142135623730951
e (nombre d'Euler) ≈ 2.718281828459045

π avec 15 décimales : 3.141592653589793


### 📊 La Droite des Nombres (Visualisation)

Imagine une ligne infinie où chaque point représente un nombre. Plus on va à droite, plus le nombre est grand.

In [1]:
import matplotlib.pyplot as plt
import numpy as np

# Créer une droite des nombres
fig, ax = plt.subplots(figsize=(12, 3))

# Ligne principale
ax.arrow(-5, 0, 10, 0, head_width=0.2, head_length=0.3, fc='black', ec='black')

# Marques pour les nombres entiers
for i in range(-4, 5):
    ax.plot([i, i], [-0.1, 0.1], 'k-', linewidth=2)
    ax.text(i, -0.4, str(i), ha='center', fontsize=12)

# Ajouter quelques nombres spéciaux
ax.plot([1.5], [0], 'ro', markersize=10)
ax.text(1.5, 0.5, '1.5', ha='center', fontsize=10, color='red')

ax.plot([-2.5], [0], 'bo', markersize=10)
ax.text(-2.5, 0.5, '-2.5', ha='center', fontsize=10, color='blue')

ax.plot([math.pi], [0], 'go', markersize=10)
ax.text(math.pi, 0.5, 'π ≈ 3.14', ha='center', fontsize=10, color='green')

ax.set_ylim(-1, 1)
ax.set_xlim(-5.5, 5.5)
ax.axis('off')
ax.set_title('La Droite des Nombres Réels', fontsize=14, fontweight='bold')

plt.tight_layout()
plt.show()

ModuleNotFoundError: No module named 'matplotlib'

**📝 Résumé des types de nombres :**

$$\mathbb{N} \subset \mathbb{Z} \subset \mathbb{Q} \subset \mathbb{R}$$

Cela signifie :
- Tous les nombres naturels sont des entiers
- Tous les entiers sont des rationnels
- Tous les rationnels sont des réels
- Mais PAS l'inverse !

---
## 2️⃣ Opérations de Base (Rappels Rapides)

Tu connais déjà ces opérations, mais rappelons-nous les règles importantes !

### ➕➖✖️➗ Les Quatre Opérations

1. **Addition** : $$5 + 3 = 8$$
2. **Soustraction** : $$7 - 2 = 5$$
3. **Multiplication** : $$4 \times 6 = 24$$
4. **Division** : $$15 \div 3 = 5$$ ou $$\frac{15}{3} = 5$$

In [ ]:
# Les quatre opérations en Python
a = 10
b = 3

addition = a + b
soustraction = a - b
multiplication = a * b
division = a / b

print(f"{a} + {b} = {addition}")
print(f"{a} - {b} = {soustraction}")
print(f"{a} × {b} = {multiplication}")
print(f"{a} ÷ {b} = {division}")

### 🎯 Priorité des Opérations (PEMDAS)

Quand tu as plusieurs opérations, dans quel ordre les faire ? Il y a une règle !

**PEMDAS** (ou **BODMAS** en français) :
1. **P**arenthèses (ce qui est entre parenthèses)
2. **E**xposants (puissances)
3. **M**ultiplication et **D**ivision (de gauche à droite)
4. **A**ddition et **S**oustraction (de gauche à droite)

**Exemple :** $$2 + 3 \times 4 = ?$$

❌ **FAUX** : $$(2 + 3) \times 4 = 5 \times 4 = 20$$

✅ **CORRECT** : $$2 + (3 \times 4) = 2 + 12 = 14$$

**Pourquoi ?** La multiplication vient AVANT l'addition dans PEMDAS !

In [ ]:
# Exemples de priorité des opérations

# Exemple 1: Multiplication avant addition
resultat1 = 2 + 3 * 4
print(f"2 + 3 × 4 = {resultat1}")  # 14, pas 20 !

# Exemple 2: Parenthèses d'abord
resultat2 = (2 + 3) * 4
print(f"(2 + 3) × 4 = {resultat2}")  # 20

# Exemple 3: Expression complexe
resultat3 = 10 - 2 * 3 + 4
print(f"\n10 - 2 × 3 + 4 = {resultat3}")
print("Étapes:")
print("  1) 2 × 3 = 6")
print("  2) 10 - 6 = 4")
print("  3) 4 + 4 = 8")

# Exemple 4: Avec parenthèses
resultat4 = (10 - 2) * (3 + 4)
print(f"\n(10 - 2) × (3 + 4) = {resultat4}")
print("Étapes:")
print("  1) 10 - 2 = 8")
print("  2) 3 + 4 = 7")
print("  3) 8 × 7 = 56")

---
## 3️⃣ Les Puissances : La Multiplication Rapide !

Imagine que tu dois multiplier un nombre par lui-même plusieurs fois. C'est fatiguant d'écrire :
$$2 \times 2 \times 2 \times 2 \times 2 = ?$$

Les **puissances** sont une façon rapide d'écrire ça !

### 💪 Qu'est-ce qu'une Puissance ?

Une puissance s'écrit : $$x^n$$

Où :
- $$x$$ s'appelle la **base**
- $$n$$ s'appelle l'**exposant**

**Définition :** $$x^n = \underbrace{x \times x \times x \times ... \times x}_{n \text{ fois}}$$

**Exemples :**
- $$2^3 = 2 \times 2 \times 2 = 8$$
- $$5^2 = 5 \times 5 = 25$$
- $$10^4 = 10 \times 10 \times 10 \times 10 = 10000$$

**Vocabulaire spécial :**
- $$x^2$$ se lit "x au carré"
- $$x^3$$ se lit "x au cube"
- $$x^n$$ se lit "x puissance n" ou "x exposant n"

In [ ]:
# Calculer des puissances en Python

# Méthode 1: Opérateur **
print("Méthode 1 : Opérateur **")
print(f"2³ = {2**3}")
print(f"5² = {5**2}")
print(f"10⁴ = {10**4}")

# Méthode 2: Fonction pow()
print("\nMéthode 2 : Fonction pow()")
print(f"2³ = {pow(2, 3)}")
print(f"5² = {pow(5, 2)}")
print(f"10⁴ = {pow(10, 4)}")

# Visualisation de la croissance
print("\nCroissance de 2ⁿ:")
for n in range(1, 11):
    print(f"2^{n} = {2**n}")

### 📏 Les Règles des Puissances

Il y a quelques règles super pratiques pour travailler avec les puissances !

#### Règle 1 : Multiplication de puissances (même base)

$$x^a \times x^b = x^{a+b}$$

**Exemple :** $$2^3 \times 2^4 = 2^{3+4} = 2^7$$

**Pourquoi ?**
$$2^3 \times 2^4 = (2 \times 2 \times 2) \times (2 \times 2 \times 2 \times 2) = 2^7$$

On a 3 "deux" puis 4 "deux" = 7 "deux" au total !

In [ ]:
# Vérification de la règle 1
base = 2
a = 3
b = 4

# Méthode longue
resultat_long = (base ** a) * (base ** b)
print(f"Méthode longue : {base}^{a} × {base}^{b} = {base**a} × {base**b} = {resultat_long}")

# Méthode avec la règle
resultat_court = base ** (a + b)
print(f"Avec la règle : {base}^{a} × {base}^{b} = {base}^{a+b} = {resultat_court}")

print(f"\nC'est pareil ! {resultat_long} = {resultat_court}")

#### Règle 2 : Division de puissances (même base)

$$\frac{x^a}{x^b} = x^{a-b}$$

**Exemple :** $$\frac{2^5}{2^2} = 2^{5-2} = 2^3$$

**Pourquoi ?**
$$\frac{2^5}{2^2} = \frac{2 \times 2 \times 2 \times 2 \times 2}{2 \times 2} = \frac{\cancel{2} \times \cancel{2} \times 2 \times 2 \times 2}{\cancel{2} \times \cancel{2}} = 2^3$$

On annule 2 "deux" en haut et en bas, il reste 3 "deux" !

In [ ]:
# Vérification de la règle 2
base = 2
a = 5
b = 2

# Méthode longue
resultat_long = (base ** a) / (base ** b)
print(f"Méthode longue : {base}^{a} ÷ {base}^{b} = {base**a} ÷ {base**b} = {resultat_long}")

# Méthode avec la règle
resultat_court = base ** (a - b)
print(f"Avec la règle : {base}^{a} ÷ {base}^{b} = {base}^{a-b} = {resultat_court}")

print(f"\nC'est pareil ! {resultat_long} = {resultat_court}")

#### Règle 3 : Puissance d'une puissance

$$(x^a)^b = x^{a \times b}$$

**Exemple :** $$(2^3)^2 = 2^{3 \times 2} = 2^6$$

**Pourquoi ?**
$$(2^3)^2 = 2^3 \times 2^3 = (2 \times 2 \times 2) \times (2 \times 2 \times 2) = 2^6$$

On multiplie $$2^3$$ par lui-même 2 fois, donc on a $$3 \times 2 = 6$$ "deux" !

In [ ]:
# Vérification de la règle 3
base = 2
a = 3
b = 2

# Méthode longue
resultat_long = (base ** a) ** b
print(f"Méthode longue : ({base}^{a})^{b} = {base**a}^{b} = {resultat_long}")

# Méthode avec la règle
resultat_court = base ** (a * b)
print(f"Avec la règle : ({base}^{a})^{b} = {base}^{a×b} = {resultat_court}")

print(f"\nC'est pareil ! {resultat_long} = {resultat_court}")

### 🔮 Puissances Spéciales

Quelques cas particuliers à connaître :

#### Puissance 0

$$x^0 = 1$$ (pour tout $$x \neq 0$$)

**Exemples :**
- $$5^0 = 1$$
- $$100^0 = 1$$
- $$(-3)^0 = 1$$

**Pourquoi ?** En utilisant la règle 2 :
$$\frac{x^3}{x^3} = 1$$ (on divise un nombre par lui-même)
$$\frac{x^3}{x^3} = x^{3-3} = x^0$$

Donc $$x^0 = 1$$ !

In [ ]:
# Puissance 0
nombres = [5, 100, -3, 0.5]

print("Tout nombre (sauf 0) à la puissance 0 égale 1 :")
for n in nombres:
    print(f"{n}^0 = {n**0}")

#### Puissance 1

$$x^1 = x$$

**Exemples :**
- $$5^1 = 5$$
- $$100^1 = 100$$

**Pourquoi ?** Multiplier un nombre par lui-même 1 fois = le nombre lui-même !

#### Puissances Négatives

$$x^{-n} = \frac{1}{x^n}$$

**Exemples :**
- $$2^{-3} = \frac{1}{2^3} = \frac{1}{8}$$
- $$5^{-2} = \frac{1}{5^2} = \frac{1}{25}$$

**Pourquoi ?** En utilisant la règle 2 :
$$\frac{x^2}{x^5} = x^{2-5} = x^{-3}$$

Mais aussi :
$$\frac{x^2}{x^5} = \frac{x \times x}{x \times x \times x \times x \times x} = \frac{1}{x^3}$$

Donc $$x^{-3} = \frac{1}{x^3}$$ !

In [ ]:
# Puissances négatives
base = 2
exposants = [-1, -2, -3, -4]

print("Puissances négatives = fractions :")
for exp in exposants:
    resultat = base ** exp
    fraction = f"1/{base**abs(exp)}"
    print(f"{base}^{exp} = {fraction} = {resultat}")

### 📊 Visualisation : Croissance Exponentielle

Les puissances grandissent TRÈS vite ! Regardons :

In [ ]:
import matplotlib.pyplot as plt
import numpy as np

# Créer des données
x = np.arange(0, 6)
y_lineaire = 2 * x  # Croissance linéaire : 2x
y_quadratique = x ** 2  # Croissance quadratique : x²
y_exponentielle = 2 ** x  # Croissance exponentielle : 2^x

# Graphique
plt.figure(figsize=(10, 6))
plt.plot(x, y_lineaire, 'o-', label='2x (linéaire)', linewidth=2)
plt.plot(x, y_quadratique, 's-', label='x² (quadratique)', linewidth=2)
plt.plot(x, y_exponentielle, '^-', label='2^x (exponentielle)', linewidth=2)

plt.xlabel('x', fontsize=12)
plt.ylabel('y', fontsize=12)
plt.title('Comparaison des Croissances', fontsize=14, fontweight='bold')
plt.legend(fontsize=11)
plt.grid(True, alpha=0.3)

# Ajouter les valeurs
for i in range(len(x)):
    plt.text(x[i], y_exponentielle[i] + 2, f"{int(y_exponentielle[i])}", ha='center', fontsize=9)

plt.tight_layout()
plt.show()

print("Remarque : La croissance exponentielle (2^x) explose très vite !")

---
## 4️⃣ Les Racines : L'Inverse des Puissances

Les racines sont le **contraire** des puissances. Si les puissances "agrandissent", les racines "rétrécissent" !

### 🌱 Qu'est-ce qu'une Racine Carrée ?

La **racine carrée** de $$x$$ s'écrit : $$\sqrt{x}$$

**Définition :** $$\sqrt{x}$$ est le nombre qui, multiplié par lui-même, donne $$x$$.

En d'autres termes : $$\sqrt{x} \times \sqrt{x} = x$$

**Exemples :**
- $$\sqrt{9} = 3$$ car $$3 \times 3 = 9$$
- $$\sqrt{16} = 4$$ car $$4 \times 4 = 16$$
- $$\sqrt{25} = 5$$ car $$5 \times 5 = 25$$
- $$\sqrt{100} = 10$$ car $$10 \times 10 = 100$$

**Question :** Quel nombre multiplié par lui-même donne 49 ? C'est $$\sqrt{49}$$ !

Réponse : $$\sqrt{49} = 7$$ car $$7 \times 7 = 49$$

In [ ]:
import math

# Calculer des racines carrées en Python
nombres = [4, 9, 16, 25, 36, 49, 64, 81, 100]

print("Racines carrées des nombres parfaits :")
for n in nombres:
    racine = math.sqrt(n)
    print(f"√{n} = {racine:.0f} (car {racine:.0f} × {racine:.0f} = {n})")

# Vérification
print("\nVérification : √9 × √9 = 9 ?")
racine_9 = math.sqrt(9)
print(f"{racine_9} × {racine_9} = {racine_9 * racine_9}")
print("Oui ! ✓")

### 🔗 Lien entre Racines et Puissances

Les racines sont en fait des **puissances fractionnaires** !

$$\sqrt{x} = x^{\frac{1}{2}} = x^{0.5}$$

**Pourquoi ?**
$$\sqrt{x} \times \sqrt{x} = x$$
$$x^{\frac{1}{2}} \times x^{\frac{1}{2}} = x^{\frac{1}{2} + \frac{1}{2}} = x^1 = x$$

Ça marche ! ✓

**Généralisation : Racine n-ième**

$$\sqrt[n]{x} = x^{\frac{1}{n}}$$

**Exemples :**
- $$\sqrt[3]{8} = 8^{\frac{1}{3}} = 2$$ (car $$2 \times 2 \times 2 = 8$$)
- $$\sqrt[4]{16} = 16^{\frac{1}{4}} = 2$$ (car $$2 \times 2 \times 2 \times 2 = 16$$)

In [ ]:
# Racines comme puissances fractionnaires

# Racine carrée
x = 25
racine_carree_methode1 = math.sqrt(x)
racine_carree_methode2 = x ** 0.5
racine_carree_methode3 = x ** (1/2)

print("Racine carrée de 25 (trois méthodes) :")
print(f"math.sqrt({x}) = {racine_carree_methode1}")
print(f"{x}^0.5 = {racine_carree_methode2}")
print(f"{x}^(1/2) = {racine_carree_methode3}")
print("Toutes les méthodes donnent le même résultat !\n")

# Racine cubique (racine 3e)
x = 27
racine_cubique = x ** (1/3)
print(f"Racine cubique de {x} : ∛{x} = {x}^(1/3) = {racine_cubique}")
print(f"Vérification : {racine_cubique:.1f} × {racine_cubique:.1f} × {racine_cubique:.1f} = {racine_cubique**3:.0f}")

# Racine 4e
x = 16
racine_4e = x ** (1/4)
print(f"\nRacine 4e de {x} : {x}^(1/4) = {racine_4e}")
print(f"Vérification : {racine_4e} × {racine_4e} × {racine_4e} × {racine_4e} = {racine_4e**4:.0f}")

### 🧮 Racines de Nombres Non-Parfaits

Qu'est-ce qui se passe si on prend la racine carrée d'un nombre comme 2 ou 3 ?

$$\sqrt{2} = 1.41421356...$$
$$\sqrt{3} = 1.73205080...$$

Ces nombres ont une infinité de décimales sans répétition. Ce sont des **nombres irrationnels** (on ne peut pas les écrire en fraction) !

In [ ]:
# Racines de nombres non-parfaits
nombres_non_parfaits = [2, 3, 5, 7, 10, 11]

print("Racines carrées de nombres non-parfaits :")
for n in nombres_non_parfaits:
    racine = math.sqrt(n)
    print(f"√{n} ≈ {racine:.10f}...")

# Vérification
print("\nVérification : (√2)² = 2 ?")
racine_2 = math.sqrt(2)
print(f"({racine_2})² = {racine_2**2}")
print("Oui (avec petite approximation due à l'ordinateur) ! ✓")

### 📊 Visualisation : Fonction Racine Carrée

Regardons à quoi ressemble la fonction racine carrée :

In [ ]:
# Graphique de y = √x
x = np.linspace(0, 25, 100)
y = np.sqrt(x)

plt.figure(figsize=(10, 6))
plt.plot(x, y, linewidth=2, color='blue')

# Marquer quelques points importants
points_x = [0, 1, 4, 9, 16, 25]
points_y = [math.sqrt(p) for p in points_x]
plt.plot(points_x, points_y, 'ro', markersize=8)

# Ajouter des annotations
for px, py in zip(points_x, points_y):
    plt.annotate(f'({px}, {py:.0f})', xy=(px, py), xytext=(px+1, py+0.3),
                fontsize=10, ha='left')

plt.xlabel('x', fontsize=12)
plt.ylabel('√x', fontsize=12)
plt.title('Fonction Racine Carrée : y = √x', fontsize=14, fontweight='bold')
plt.grid(True, alpha=0.3)
plt.axhline(y=0, color='k', linewidth=0.5)
plt.axvline(x=0, color='k', linewidth=0.5)

plt.tight_layout()
plt.show()

print("Remarque : La racine carrée grandit de moins en moins vite !")

---
## 5️⃣ Introduction aux Variables : Les Boîtes Magiques

Jusqu'ici, on a travaillé avec des nombres fixes. Mais parfois, on veut parler d'un nombre **sans savoir exactement lequel** !

C'est là que les **variables** entrent en jeu !

### 📦 Qu'est-ce qu'une Variable ?

Une **variable** est comme une **boîte** qui peut contenir un nombre (ou autre chose).

On utilise généralement des lettres comme $$x$$, $$y$$, $$a$$, $$b$$ pour représenter les variables.

**Exemple :**
- $$x = 5$$ signifie "la boîte $$x$$ contient le nombre 5"
- $$y = 10$$ signifie "la boîte $$y$$ contient le nombre 10"

**Pourquoi c'est utile ?**
- Pour écrire des formules générales qui fonctionnent avec n'importe quel nombre
- Pour résoudre des problèmes où on cherche un nombre inconnu

In [ ]:
# Variables en Python
x = 5
y = 10

print(f"La variable x contient : {x}")
print(f"La variable y contient : {y}")

# On peut faire des opérations avec les variables
somme = x + y
produit = x * y

print(f"\nx + y = {x} + {y} = {somme}")
print(f"x × y = {x} × {y} = {produit}")

# On peut changer la valeur d'une variable
print("\nChangeons x de 5 à 7 :")
x = 7
print(f"Maintenant x = {x}")
print(f"Et x + y = {x} + {y} = {x + y}")

### 🧮 Expressions Algébriques

Une **expression algébrique** est une combinaison de nombres, variables et opérations.

**Exemples :**
- $$2x + 3$$
- $$x^2 - 5$$
- $$3x + 2y - 7$$

**Comment évaluer une expression ?**

Il suffit de **remplacer** la variable par un nombre !

**Exemple :** Si $$x = 4$$, calcule $$2x + 3$$

$$2x + 3 = 2 \times 4 + 3 = 8 + 3 = 11$$

In [ ]:
# Évaluer des expressions algébriques

# Expression 1: 2x + 3
def expression1(x):
    return 2*x + 3

# Expression 2: x² - 5
def expression2(x):
    return x**2 - 5

# Expression 3: 3x + 2y - 7
def expression3(x, y):
    return 3*x + 2*y - 7

# Tester avec différentes valeurs
print("Expression 1 : 2x + 3")
for x in [1, 2, 3, 4, 5]:
    print(f"  x = {x} → 2×{x} + 3 = {expression1(x)}")

print("\nExpression 2 : x² - 5")
for x in [1, 2, 3, 4, 5]:
    print(f"  x = {x} → {x}² - 5 = {expression2(x)}")

print("\nExpression 3 : 3x + 2y - 7")
x, y = 2, 3
print(f"  x = {x}, y = {y} → 3×{x} + 2×{y} - 7 = {expression3(x, y)}")

---
## 6️⃣ Les Équations Simples : Trouver le Nombre Mystère

Une **équation** est comme une balance en équilibre. Les deux côtés doivent être égaux !

**Résoudre une équation**, c'est trouver la valeur de $$x$$ qui rend l'égalité vraie.

### ⚖️ Équations de Type 1 : $$x + a = b$$

**Exemple :** $$x + 5 = 12$$

**Question :** Quel nombre + 5 donne 12 ?

**Méthode :** Pour isoler $$x$$, on fait la même chose des deux côtés de l'équation.

$$x + 5 = 12$$
$$x + 5 - 5 = 12 - 5$$ (on enlève 5 des deux côtés)
$$x = 7$$

**Vérification :** $$7 + 5 = 12$$ ✓

In [1]:
# Résolution d'équations de type x + a = b

# Équation : x + 5 = 12
print("Équation : x + 5 = 12")
print("Étapes :")
print("  1) x + 5 = 12")
print("  2) x + 5 - 5 = 12 - 5  (on enlève 5 des deux côtés)")
print("  3) x = 7")

x = 7
print(f"\nVérification : {x} + 5 = {x + 5}")
print("C'est bien égal à 12 ! ✓")

# Autre exemple : x + 3 = 10
print("\n" + "="*40)
print("Équation : x + 3 = 10")
print("Étapes :")
print("  1) x + 3 = 10")
print("  2) x + 3 - 3 = 10 - 3")
print("  3) x = 7")

x = 7
print(f"\nVérification : {x} + 3 = {x + 3}")
print("C'est bien égal à 10 ! ✓")

Équation : x + 5 = 12
Étapes :
  1) x + 5 = 12
  2) x + 5 - 5 = 12 - 5  (on enlève 5 des deux côtés)
  3) x = 7

Vérification : 7 + 5 = 12
C'est bien égal à 12 ! ✓

Équation : x + 3 = 10
Étapes :
  1) x + 3 = 10
  2) x + 3 - 3 = 10 - 3
  3) x = 7

Vérification : 7 + 3 = 10
C'est bien égal à 10 ! ✓


### ✖️ Équations de Type 2 : $$ax = b$$

**Exemple :** $$2x = 10$$

**Question :** Quel nombre × 2 donne 10 ?

**Méthode :** On divise les deux côtés par 2.

$$2x = 10$$
$$\frac{2x}{2} = \frac{10}{2}$$ (on divise par 2 des deux côtés)
$$x = 5$$

**Vérification :** $$2 \times 5 = 10$$ ✓

In [ ]:
# Résolution d'équations de type ax = b

# Équation : 2x = 10
print("Équation : 2x = 10")
print("Étapes :")
print("  1) 2x = 10")
print("  2) 2x ÷ 2 = 10 ÷ 2  (on divise par 2 des deux côtés)")
print("  3) x = 5")

x = 5
print(f"\nVérification : 2 × {x} = {2 * x}")
print("C'est bien égal à 10 ! ✓")

# Autre exemple : 3x = 15
print("\n" + "="*40)
print("Équation : 3x = 15")
print("Étapes :")
print("  1) 3x = 15")
print("  2) 3x ÷ 3 = 15 ÷ 3")
print("  3) x = 5")

x = 5
print(f"\nVérification : 3 × {x} = {3 * x}")
print("C'est bien égal à 15 ! ✓")

### 🎯 Équations de Type 3 : $$ax + b = c$$

**Exemple :** $$3x + 2 = 11$$

**Méthode :** On combine les deux techniques précédentes.

**Étape 1 :** Enlever le nombre seul ($$+2$$)
$$3x + 2 = 11$$
$$3x + 2 - 2 = 11 - 2$$
$$3x = 9$$

**Étape 2 :** Diviser par le coefficient de $$x$$ ($$3$$)
$$\frac{3x}{3} = \frac{9}{3}$$
$$x = 3$$

**Vérification :** $$3 \times 3 + 2 = 9 + 2 = 11$$ ✓

In [ ]:
# Résolution d'équations de type ax + b = c

# Équation : 3x + 2 = 11
print("Équation : 3x + 2 = 11")
print("\nÉtape 1 : Enlever +2 des deux côtés")
print("  3x + 2 - 2 = 11 - 2")
print("  3x = 9")

print("\nÉtape 2 : Diviser par 3 des deux côtés")
print("  3x ÷ 3 = 9 ÷ 3")
print("  x = 3")

x = 3
print(f"\nVérification : 3 × {x} + 2 = {3*x + 2}")
print("C'est bien égal à 11 ! ✓")

# Autre exemple : 2x - 5 = 7
print("\n" + "="*40)
print("Équation : 2x - 5 = 7")
print("\nÉtape 1 : Ajouter 5 des deux côtés")
print("  2x - 5 + 5 = 7 + 5")
print("  2x = 12")

print("\nÉtape 2 : Diviser par 2 des deux côtés")
print("  2x ÷ 2 = 12 ÷ 2")
print("  x = 6")

x = 6
print(f"\nVérification : 2 × {x} - 5 = {2*x - 5}")
print("C'est bien égal à 7 ! ✓")

### 🔧 Outil : Résolveur d'Équations Interactif

In [ ]:
def resoudre_equation_simple(a, b, c):
    """
    Résout une équation de la forme : ax + b = c
    
    Exemples:
    - Pour x + 5 = 12 → a=1, b=5, c=12
    - Pour 2x = 10 → a=2, b=0, c=10
    - Pour 3x + 2 = 11 → a=3, b=2, c=11
    """
    print(f"Équation : {a}x + {b} = {c}")
    print("\nÉtapes de résolution :")
    
    # Étape 1 : Soustraire b
    if b != 0:
        print(f"  1) {a}x + {b} - {b} = {c} - {b}")
        nouveau_c = c - b
        print(f"     {a}x = {nouveau_c}")
    else:
        nouveau_c = c
        print(f"  1) {a}x = {c}")
    
    # Étape 2 : Diviser par a
    print(f"  2) {a}x ÷ {a} = {nouveau_c} ÷ {a}")
    x = nouveau_c / a
    print(f"     x = {x}")
    
    # Vérification
    verification = a * x + b
    print(f"\nVérification : {a} × {x} + {b} = {verification}")
    if abs(verification - c) < 0.0001:  # Tolérance pour les erreurs d'arrondi
        print("C'est correct ! ✓")
    else:
        print("Erreur dans le calcul ! ✗")
    
    return x

# Exemples d'utilisation
print("Exemple 1:")
resoudre_equation_simple(a=1, b=5, c=12)  # x + 5 = 12

print("\n" + "="*50 + "\n")
print("Exemple 2:")
resoudre_equation_simple(a=2, b=0, c=10)  # 2x = 10

print("\n" + "="*50 + "\n")
print("Exemple 3:")
resoudre_equation_simple(a=3, b=2, c=11)  # 3x + 2 = 11

---
## 🎉 Félicitations !

Tu as terminé ce cours sur les bases des mathématiques !

**Ce que tu as appris :**
- ✅ Les différents types de nombres (naturels, entiers, rationnels, réels)
- ✅ Les puissances et leurs règles
- ✅ Les racines (l'inverse des puissances)
- ✅ Les variables et expressions algébriques
- ✅ Comment résoudre des équations simples

---

## ✏️ Prochaine étape : Exercices !

**Maintenant, entraîne-toi avec les exercices :**

📁 `envs/phase_1_math/exercices_01_bases.ipynb`

**Ensuite, vérifie tes réponses :**

📁 `envs/phase_1_math/solutions_01_bases.ipynb`

---

**Conseils :**
- Essaie **VRAIMENT** chaque exercice avant de regarder la solution
- Si tu bloques > 15 min, regarde un indice puis réessaie
- Refais les exercices ratés pour bien consolider

**N'oublie pas :** Les maths, c'est comme un muscle. Plus tu t'entraînes, plus tu deviens fort ! 💪

Continue ton apprentissage avec le cours suivant : `Maths_02_Fonctions.ipynb` 🚀